libfuse/quarantine_darwin.go

```diff
@@ -0,0 +1,135 @@
+// com.apple.quarantine, but returns fuse.ENOTSUP for all other xattrs.
+func NewQuarantineXattrHandler(node libkbfs.Node, folder *Folder) XattrHandler {
+	return &QuarantineXattrHandler{node: node, folder: folder}
+}
+
+var _ XattrHandler = (*QuarantineXattrHandler)(nil)
+
+// Getxattr implements the fs.NodeGetxattrer interface.
+func (h *QuarantineXattrHandler) Getxattr(ctx context.Context,
+	req *fuse.GetxattrRequest, resp *fuse.GetxattrResponse) (err error) {
+	ctx = h.folder.fs.config.MaybeStartTrace(ctx, "QuarantineXattrHandler.Getxattr",
+		fmt.Sprintf("%s %s", h.node.GetBasename(), req.Name))
+	defer func() { h.folder.fs.config.MaybeFinishTrace(ctx, err) }()
+
+	h.folder.fs.log.CDebugf(ctx, "QuarantineXattrHandler Getxattr %s %s", h.node.GetBasename(), req.Name)
+	defer func() { err = h.folder.processError(ctx, libkbfs.ReadMode, err) }()
+
+	if req.Name != quarantineXattrName {
+		// The request is not about quarantine. Let the OS fallback to ._ file
+		// based method.
+		return fuse.ENOTSUP
+	}
+
+	if h.bytesSet {
+		if h.bytes == nil {
+			return fuse.ENOATTR
+		}
+		resp.Xattr = make([]byte, len(h.bytes))
+		copy(resp.Xattr, h.bytes)
+		return nil
+	}
+
+	// Stat on the node to get the Mtime.
+
+	// This fits in situation 1 as described in libkbfs/delayed_cancellation.go
+	if err = libkbfs.EnableDelayedCancellationWithGracePeriod(
+		ctx, h.folder.fs.config.DelayedCancellationGracePeriod()); err != nil {
+		return err
+	}
+	de, err := h.folder.fs.config.KBFSOps().Stat(ctx, h.node)
+	if err != nil {
+		if _, ok := err.(libkbfs.NoSuchNameError); ok {
+			// The node is not found, so just return ENOTSUP
+			return fuse.ENOTSUP
+		}
+		return err
+	}
+
+	resp.Xattr = makeQuarantine(time.Unix(0, de.Mtime))
+	return nil
+}
+
+// Setxattr implements the fs.NodeSetxattrer interface.
+func (h *QuarantineXattrHandler) Setxattr(ctx context.Context,
+	req *fuse.SetxattrRequest) (err error) {
+	ctx = h.folder.fs.config.MaybeStartTrace(ctx, "QuarantineXattrHandler.Setxattr",
+		fmt.Sprintf("%s %s", h.node.GetBasename(), req.Name))
+	defer func() { h.folder.fs.config.MaybeFinishTrace(ctx, err) }()
+
+	h.folder.fs.log.CDebugf(ctx, "QuarantineXattrHandler Setxattr %s", req.Name)
+	defer func() { err = h.folder.processError(ctx, libkbfs.ReadMode, err) }()
+
+	if req.Name != quarantineXattrName {
+		// The request is not about quarantine. Let the OS fallback to ._ file
+		// based method.
+		return fuse.ENOTSUP
+	}
+
+	// TODO persist to local db, and clear if block ID changes
+	h.bytes = make([]byte, len(req.Xattr))
+	copy(h.bytes, req.Xattr)
+	h.bytesSet = true
+
+	return nil
+}
+
+// Removexattr implements the fs.NodeRemovexattrer interface.
+func (h *QuarantineXattrHandler) Removexattr(ctx context.Context, req *fuse.RemovexattrRequest) error {
+	if req.Name != quarantineXattrName {
+		// The request is not about quarantine. Let the OS fallback to ._ file
+		// based method.
+		return fuse.ENOTSUP
+	}
+
+	h.bytes = nil
+	h.bytesSet = true
```

Comment:
Should this be false?

Reply:
I think this should be `true`, so next time Get happens, we'd return `ENOATTR`. If a syscall came in to explicitly remove that attr, we should remember it.